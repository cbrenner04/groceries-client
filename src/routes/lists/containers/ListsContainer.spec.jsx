import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { toast } from 'react-toastify';

import ListsContainer from './ListsContainer';
import axios from '../../../utils/api';

jest.mock('react-toastify', () => ({
  toast: jest.fn(),
}));

describe('ListsContainer', () => {
  let props;
  const renderListsContainer = (props) => {
    const history = createMemoryHistory();
    return render(
      <Router history={history}>
        <ListsContainer {...props} />
      </Router>,
    );
  };

  beforeEach(() => {
    jest.useFakeTimers();
    props = {
      history: {
        push: jest.fn(),
        replace: jest.fn(),
        location: {
          pathname: '/lists',
        },
      },
      userId: 'id1',
      pendingLists: [
        {
          id: 'id1',
          name: 'foo',
          type: 'GroceryList',
          created_at: new Date('05/31/2020').toISOString(),
          completed: false,
          users_list_id: 'id1',
          owner_id: 'id1',
          refreshed: false,
        },
      ],
      completedLists: [
        {
          id: 'id2',
          name: 'bar',
          type: 'BookList',
          created_at: new Date('05/31/2020').toISOString(),
          completed: true,
          users_list_id: 'id2',
          owner_id: 'id1',
          refreshed: false,
        },
        {
          id: 'id4',
          name: 'bar',
          type: 'BookList',
          created_at: new Date('05/31/2020').toISOString(),
          completed: true,
          users_list_id: 'id4',
          owner_id: 'id2',
          refreshed: false,
        },
      ],
      incompleteLists: [
        {
          id: 'id3',
          name: 'baz',
          type: 'MusicList',
          created_at: new Date('05/31/2020').toISOString(),
          completed: false,
          users_list_id: 'id3',
          owner_id: 'id1',
          refreshed: false,
        },
        {
          id: 'id5',
          name: 'foobar',
          type: 'ToDoList',
          created_at: new Date('05/31/2020').toISOString(),
          completed: false,
          users_list_id: 'id5',
          owner_id: 'id1',
          refreshed: false,
        },
      ],
      currentUserPermissions: {
        id1: 'write',
        id2: 'write',
        id3: 'write',
        id4: 'read',
        id5: 'read',
      },
    };
  });

  it('renders', () => {
    const { container } = renderListsContainer(props);

    expect(container).toMatchSnapshot();
  });

  it('updates via polling when different data is returned', async () => {
    axios.get = jest
      .fn()
      .mockResolvedValueOnce({
        data: {
          current_user_id: 'id1',
          accepted_lists: {
            completed_lists: [
              {
                id: 'id1',
                users_list_id: 'id1',
                name: 'foo',
                user_id: 'id1',
                type: 'GroceryList',
                created_at: new Date('05/31/2020').toISOString(),
                completed: true,
                refreshed: false,
                owner_id: 'id1',
              },
            ],
            not_completed_lists: [
              {
                id: 'id2',
                users_list_id: 'id2',
                name: 'bar',
                user_id: 'id1',
                type: 'BookList',
                created_at: new Date('05/31/2020').toISOString(),
                completed: false,
                refreshed: false,
                owner_id: 'id1',
              },
            ],
          },
          pending_lists: [
            {
              id: 'id3',
              users_list_id: 'id3',
              name: 'foo',
              user_id: 'id1',
              type: 'GroceryList',
              created_at: new Date('05/31/2020').toISOString(),
              completed: false,
              refreshed: false,
              owner_id: 'id2',
            },
          ],
          current_list_permissions: {
            id1: 'write',
            id2: 'write',
            id3: 'write',
          },
        },
      })
      .mockResolvedValueOnce({
        data: {
          current_user_id: 'id1',
          accepted_lists: {
            completed_lists: [
              {
                id: 'id1',
                users_list_id: 'id1',
                name: 'foo',
                user_id: 'id1',
                type: 'GroceryList',
                created_at: new Date('05/31/2020').toISOString(),
                completed: true,
                refreshed: false,
                owner_id: 'id1',
              },
              {
                id: 'id3',
                users_list_id: 'id3',
                name: 'foo',
                user_id: 'id1',
                type: 'GroceryList',
                created_at: new Date('05/31/2020').toISOString(),
                completed: false,
                refreshed: false,
                owner_id: 'id2',
              },
            ],
            not_completed_lists: [
              {
                id: 'id2',
                users_list_id: 'id2',
                name: 'bar',
                user_id: 'id1',
                type: 'BookList',
                created_at: new Date('05/31/2020').toISOString(),
                completed: false,
                refreshed: false,
                owner_id: 'id1',
              },
            ],
          },
          pending_lists: [],
          current_list_permissions: {
            id1: 'write',
            id2: 'write',
            id3: 'write',
          },
        },
      });

    const { getByTestId } = renderListsContainer(props);

    await waitFor(() => expect(axios.get).toHaveBeenCalledTimes(1));

    expect(getByTestId('list-id3')).toHaveAttribute('data-test-class', 'pending-list');

    await waitFor(() => expect(axios.get).toHaveBeenCalledTimes(2));

    expect(getByTestId('list-id3')).toHaveAttribute('data-test-class', 'completed-list');
  });

  it('does not update via polling when different data is not returned', async () => {
    axios.get = jest.fn().mockResolvedValue({
      data: {
        current_user_id: 'id1',
        accepted_lists: {
          completed_lists: [
            {
              id: 'id1',
              users_list_id: 'id1',
              name: 'foo',
              user_id: 'id1',
              type: 'GroceryList',
              created_at: new Date('05/31/2020').toISOString(),
              completed: true,
              refreshed: false,
              owner_id: 'id1',
            },
          ],
          not_completed_lists: [
            {
              id: 'id2',
              users_list_id: 'id2',
              name: 'bar',
              user_id: 'id1',
              type: 'BookList',
              created_at: new Date('05/31/2020').toISOString(),
              completed: false,
              refreshed: false,
              owner_id: 'id1',
            },
          ],
        },
        pending_lists: [
          {
            id: 'id3',
            users_list_id: 'id3',
            name: 'foo',
            user_id: 'id1',
            type: 'GroceryList',
            created_at: new Date('05/31/2020').toISOString(),
            completed: false,
            refreshed: false,
            owner_id: 'id2',
          },
        ],
        current_list_permissions: {
          id1: 'write',
          id2: 'write',
          id3: 'write',
        },
      },
    });

    const { getByTestId } = renderListsContainer(props);

    await waitFor(() => expect(axios.get).toHaveBeenCalledTimes(1));

    expect(getByTestId('list-id3')).toHaveAttribute('data-test-class', 'pending-list');

    await waitFor(() => expect(axios.get).toHaveBeenCalledTimes(2));

    expect(getByTestId('list-id3')).toHaveAttribute('data-test-class', 'pending-list');
  });

  it('does not render pending lists when they do not exist', () => {
    props.pendingLists = [];
    const { container, queryByText } = renderListsContainer(props);

    expect(container).toMatchSnapshot();
    expect(queryByText('Pending')).toBeNull();
  });

  it('creates list on form submit', async () => {
    axios.post = jest.fn().mockResolvedValue({
      data: {
        id: 'id6',
        name: 'new list',
        type: 'BookList',
        created_at: new Date('05/31/2020').toISOString(),
        owner_id: 'id1',
        completed: false,
        refreshed: false,
        users_list_id: 'id9',
      },
    });
    const { getByLabelText, getByTestId, getByText } = renderListsContainer(props);

    fireEvent.change(getByLabelText('Name'), { target: { value: 'new list' } });
    fireEvent.change(getByLabelText('Type'), { target: { value: 'BookList' } });
    fireEvent.click(getByText('Create List'));
    await waitFor(() => expect(axios.post).toHaveBeenCalledTimes(1));

    expect(toast).toHaveBeenCalledWith('List successfully added.', { type: 'info' });
    expect(getByTestId('list-id6')).toHaveTextContent('new list');
  });

  it('redirects to login when submit response is 401', async () => {
    axios.post = jest.fn().mockRejectedValue({ response: { status: 401 } });
    const { getByLabelText, getByText } = renderListsContainer(props);

    fireEvent.change(getByLabelText('Name'), { target: { value: 'new list' } });
    fireEvent.change(getByLabelText('Type'), { target: { value: 'BookList' } });
    fireEvent.click(getByText('Create List'));
    await waitFor(() => expect(axios.post).toHaveBeenCalledTimes(1));

    expect(toast).toHaveBeenCalledWith('You must sign in', { type: 'error' });
    expect(props.history.push).toHaveBeenCalledWith('/users/sign_in');
  });

  it('shows errors when submission fails', async () => {
    axios.post = jest.fn().mockRejectedValue({ response: { status: 400, data: { foo: 'bar', foobar: 'foobaz' } } });
    const { getByLabelText, getByText } = renderListsContainer(props);

    fireEvent.change(getByLabelText('Name'), { target: { value: 'new list' } });
    fireEvent.change(getByLabelText('Type'), { target: { value: 'BookList' } });
    fireEvent.click(getByText('Create List'));
    await waitFor(() => expect(axios.post).toHaveBeenCalledTimes(1));

    expect(toast).toHaveBeenCalledWith('foo bar and foobar foobaz', { type: 'error' });
  });

  it('shows errors when request fails', async () => {
    axios.post = jest.fn().mockRejectedValue({ request: 'failed to send request' });
    const { getByLabelText, getByText } = renderListsContainer(props);

    fireEvent.change(getByLabelText('Name'), { target: { value: 'new list' } });
    fireEvent.change(getByLabelText('Type'), { target: { value: 'BookList' } });
    fireEvent.click(getByText('Create List'));
    await waitFor(() => expect(axios.post).toHaveBeenCalledTimes(1));

    expect(toast).toHaveBeenCalledWith('Something went wrong', { type: 'error' });
  });

  it('shows errors when unknown error occurs', async () => {
    axios.post = jest.fn().mockRejectedValue({ message: 'failed to send request' });
    const { getByLabelText, getByText } = renderListsContainer(props);

    fireEvent.change(getByLabelText('Name'), { target: { value: 'new list' } });
    fireEvent.change(getByLabelText('Type'), { target: { value: 'BookList' } });
    fireEvent.click(getByText('Create List'));
    await waitFor(() => expect(axios.post).toHaveBeenCalledTimes(1));

    expect(toast).toHaveBeenCalledWith('failed to send request', { type: 'error' });
  });
});
