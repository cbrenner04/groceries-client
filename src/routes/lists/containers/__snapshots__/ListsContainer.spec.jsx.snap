// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListsContainer does not render pending lists when they do not exist 1`] = `
<div>
  <h1>
    Lists
  </h1>
  <button
    aria-controls="form-collapse"
    aria-expanded="false"
    class="btn btn-link"
    type="button"
  >
    Add List
  </button>
  <form
    autocomplete="off"
    class="collapse"
    id="form-collapse"
  >
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="name"
      >
        Name
      </label>
      <input
        class="form-control"
        id="name"
        name="name"
        placeholder="My super cool list"
        type="text"
        value=""
      />
      
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="type"
      >
        Type
      </label>
      <select
        class="form-control"
        id="type"
        name="type"
      >
        <option
          value="BookList"
        >
          books
        </option>
        <option
          value="GroceryList"
        >
          groceries
        </option>
        <option
          value="MusicList"
        >
          music
        </option>
        <option
          value="SimpleList"
        >
          simple
        </option>
        <option
          value="ToDoList"
        >
          to-do
        </option>
      </select>
      
    </div>
    <div
      class="d-grid gap-2 mt-3"
    >
      <button
        class="btn btn-success"
        type="submit"
      >
        Create List
      </button>
      <button
        class="btn btn-link"
        type="button"
      >
        Collapse Form
      </button>
    </div>
  </form>
  <hr
    class="mb-4"
  />
  <div
    class="mb-4"
  >
    <div
      class="clearfix"
    >
      <div
        class="mx-auto float-start"
      >
        <div
          class="m-0 row"
        >
          <h2
            class="text-capitalize pe-0 w-auto"
          >
            Incomplete
          </h2>
          <i
            class="far fa-question-circle float-end text-secondary ps-0 w-auto"
            data-test-id="Incomplete-popover"
          />
        </div>
      </div>
      <button
        class="mx-auto float-end btn btn-link"
        type="button"
      >
        
         Select
      </button>
    </div>
    <div
      class="list-group"
    >
      <div
        class="accepted-list list-list-group-item list-group-item"
        data-test-class="incomplete-list"
        data-test-id="list-id3"
      >
        <div
          class="row"
        >
          <div
            class="col-sm-12 col-12"
          >
            <div
              class=" pt-1 row"
            >
              <div
                class="col-lg-6"
              >
                <a
                  class="router-link"
                  href="/lists/id3"
                >
                  <h5
                    class="mb-1"
                  >
                    <i
                      class="fa fa-music text-secondary me-3"
                    />
                    baz
                  </h5>
                </a>
              </div>
              <div
                class="col-lg-4"
              >
                <small
                  class="text-muted"
                >
                  May 31 2020, 12:00:00 am
                </small>
              </div>
              <div
                class=" pe-lg-3 col-lg-2"
              >
                <div
                  class="float-end btn-group"
                  role="group"
                >
                  <button
                    class="p-0 me-3 btn btn-link"
                    data-test-id="incomplete-list-complete"
                    type="button"
                  >
                    <i
                      class="fa fa-check fa-2x text-success"
                    />
                  </button>
                  <a
                    class="p-0 me-3 list-button-enabled"
                    data-test-id="incomplete-list-share"
                    href="/lists/id3/users_lists"
                  >
                    <i
                      class="fa fa-users fa-2x text-primary"
                    />
                  </a>
                  <a
                    class="p-0 mr-3 list-button-enabled"
                    data-test-id="incomplete-list-edit"
                    href="/lists/id3/edit"
                  >
                    <i
                      class="fa fa-edit fa-2x text-warning"
                    />
                  </a>
                  <button
                    class="p-0 btn btn-link"
                    data-test-id="incomplete-list-trash"
                    type="button"
                  >
                    <i
                      class="fa fa-trash fa-2x text-danger"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="accepted-list list-list-group-item list-group-item"
        data-test-class="incomplete-list"
        data-test-id="list-id5"
      >
        <div
          class="row"
        >
          <div
            class="col-sm-12 col-12"
          >
            <div
              class=" pt-1 row"
            >
              <div
                class="col-lg-6"
              >
                <a
                  class="router-link"
                  href="/lists/id5"
                >
                  <h5
                    class="mb-1"
                  >
                    <i
                      class="fa fa-list text-secondary me-3"
                    />
                    foobar
                  </h5>
                </a>
              </div>
              <div
                class="col-lg-4"
              >
                <small
                  class="text-muted"
                >
                  May 31 2020, 12:00:00 am
                </small>
              </div>
              <div
                class=" pe-lg-3 col-lg-2"
              >
                <div
                  class="float-end btn-group"
                  role="group"
                >
                  <button
                    class="p-0 me-3 btn btn-link"
                    data-test-id="incomplete-list-complete"
                    type="button"
                  >
                    <i
                      class="fa fa-check fa-2x text-success"
                    />
                  </button>
                  <a
                    class="p-0 me-3 list-button-disabled"
                    data-test-id="incomplete-list-share"
                    disabled=""
                    href="/lists/id5/users_lists"
                  >
                    <i
                      class="fa fa-users fa-2x text-primary"
                    />
                  </a>
                  <a
                    class="p-0 mr-3 list-button-enabled"
                    data-test-id="incomplete-list-edit"
                    href="/lists/id5/edit"
                  >
                    <i
                      class="fa fa-edit fa-2x text-warning"
                    />
                  </a>
                  <button
                    class="p-0 btn btn-link"
                    data-test-id="incomplete-list-trash"
                    type="button"
                  >
                    <i
                      class="fa fa-trash fa-2x text-danger"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="mb-4"
  >
    <div
      class="clearfix"
    >
      <div
        class="mx-auto float-start"
      >
        <div
          class="m-0 row"
        >
          <h2
            class="text-capitalize pe-0 w-auto"
          >
            Completed
          </h2>
          <i
            class="far fa-question-circle float-end text-secondary ps-0 w-auto"
            data-test-id="Completed-popover"
          />
        </div>
      </div>
      <button
        class="mx-auto float-end btn btn-link"
        type="button"
      >
        
         Select
      </button>
    </div>
    <div
      class="list-group"
    >
      <div
        class="accepted-list list-list-group-item list-group-item"
        data-test-class="completed-list"
        data-test-id="list-id2"
      >
        <div
          class="row"
        >
          <div
            class="col-sm-12 col-12"
          >
            <div
              class=" pt-1 row"
            >
              <div
                class="col-lg-6"
              >
                <a
                  class="router-link"
                  href="/lists/id2"
                >
                  <h5
                    class="mb-1"
                  >
                    <i
                      class="fa fa-book text-secondary me-3"
                    />
                    bar
                  </h5>
                </a>
              </div>
              <div
                class="col-lg-4"
              >
                <small
                  class="text-muted"
                >
                  May 31 2020, 12:00:00 am
                </small>
              </div>
              <div
                class=" pe-lg-3 col-lg-2"
              >
                <div
                  class="float-end btn-group"
                  role="group"
                >
                  <button
                    class="p-0 me-3 list-button-enabled btn btn-link"
                    data-test-id="complete-list-refresh"
                    type="button"
                  >
                    <i
                      class="fa fa-redo fa-2x text-primary"
                    />
                  </button>
                  <button
                    class="p-0 btn btn-link"
                    data-test-id="complete-list-trash"
                    type="button"
                  >
                    <i
                      class="fa fa-trash fa-2x text-danger"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="accepted-list list-list-group-item list-group-item"
        data-test-class="completed-list"
        data-test-id="list-id4"
      >
        <div
          class="row"
        >
          <div
            class="col-sm-12 col-12"
          >
            <div
              class=" pt-1 row"
            >
              <div
                class="col-lg-6"
              >
                <a
                  class="router-link"
                  href="/lists/id4"
                >
                  <h5
                    class="mb-1"
                  >
                    <i
                      class="fa fa-book text-secondary me-3"
                    />
                    bar
                  </h5>
                </a>
              </div>
              <div
                class="col-lg-4"
              >
                <small
                  class="text-muted"
                >
                  May 31 2020, 12:00:00 am
                </small>
              </div>
              <div
                class=" pe-lg-3 col-lg-2"
              >
                <div
                  class="float-end btn-group"
                  role="group"
                >
                  <button
                    class="p-0 me-3 list-button-disabled btn btn-link"
                    data-test-id="complete-list-refresh"
                    disabled=""
                    type="button"
                  >
                    <i
                      class="fa fa-redo fa-2x text-primary"
                    />
                  </button>
                  <button
                    class="p-0 btn btn-link"
                    data-test-id="complete-list-trash"
                    type="button"
                  >
                    <i
                      class="fa fa-trash fa-2x text-danger"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ListsContainer renders 1`] = `
<div>
  <h1>
    Lists
  </h1>
  <button
    aria-controls="form-collapse"
    aria-expanded="false"
    class="btn btn-link"
    type="button"
  >
    Add List
  </button>
  <form
    autocomplete="off"
    class="collapse"
    id="form-collapse"
  >
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="name"
      >
        Name
      </label>
      <input
        class="form-control"
        id="name"
        name="name"
        placeholder="My super cool list"
        type="text"
        value=""
      />
      
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="type"
      >
        Type
      </label>
      <select
        class="form-control"
        id="type"
        name="type"
      >
        <option
          value="BookList"
        >
          books
        </option>
        <option
          value="GroceryList"
        >
          groceries
        </option>
        <option
          value="MusicList"
        >
          music
        </option>
        <option
          value="SimpleList"
        >
          simple
        </option>
        <option
          value="ToDoList"
        >
          to-do
        </option>
      </select>
      
    </div>
    <div
      class="d-grid gap-2 mt-3"
    >
      <button
        class="btn btn-success"
        type="submit"
      >
        Create List
      </button>
      <button
        class="btn btn-link"
        type="button"
      >
        Collapse Form
      </button>
    </div>
  </form>
  <hr
    class="mb-4"
  />
  <div
    class="mb-4"
  >
    <div
      class="clearfix"
    >
      <div
        class="mx-auto float-start"
      >
        <div
          class="m-0 row"
        >
          <h2
            class="text-capitalize pe-0 w-auto"
          >
            Pending
          </h2>
          <i
            class="far fa-question-circle float-end text-secondary ps-0 w-auto"
            data-test-id="Pending-popover"
          />
        </div>
      </div>
      <button
        class="mx-auto float-end btn btn-link"
        type="button"
      >
        
         Select
      </button>
    </div>
    <div
      class="list-group"
    >
      <div
        class="pending-list list-list-group-item list-group-item"
        data-test-class="pending-list"
        data-test-id="list-id1"
      >
        <div
          class="row"
        >
          <div
            class="col-sm-12 col-12"
          >
            <div
              class=" pt-1 row"
            >
              <div
                class="col-lg-6"
              >
                <h5
                  class="mb-1"
                >
                  <i
                    class="fa fa-shopping-bag text-secondary me-3"
                  />
                  foo
                </h5>
              </div>
              <div
                class="col-lg-4"
              >
                <small
                  class="text-muted"
                >
                  May 31 2020, 12:00:00 am
                </small>
              </div>
              <div
                class=" pe-lg-3 col-lg-2"
              >
                <div
                  class="float-end btn-group"
                  role="group"
                >
                  <button
                    class="p-0 me-3 btn btn-link"
                    data-test-id="pending-list-accept"
                    type="button"
                  >
                    <i
                      class="fa fa-check fa-2x text-success"
                    />
                  </button>
                  <button
                    class="p-0 btn btn-link"
                    data-test-id="pending-list-trash"
                    type="button"
                  >
                    <i
                      class="fa fa-trash fa-2x text-danger"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="mb-4"
  >
    <div
      class="clearfix"
    >
      <div
        class="mx-auto float-start"
      >
        <div
          class="m-0 row"
        >
          <h2
            class="text-capitalize pe-0 w-auto"
          >
            Incomplete
          </h2>
          <i
            class="far fa-question-circle float-end text-secondary ps-0 w-auto"
            data-test-id="Incomplete-popover"
          />
        </div>
      </div>
      <button
        class="mx-auto float-end btn btn-link"
        type="button"
      >
        
         Select
      </button>
    </div>
    <div
      class="list-group"
    >
      <div
        class="accepted-list list-list-group-item list-group-item"
        data-test-class="incomplete-list"
        data-test-id="list-id3"
      >
        <div
          class="row"
        >
          <div
            class="col-sm-12 col-12"
          >
            <div
              class=" pt-1 row"
            >
              <div
                class="col-lg-6"
              >
                <a
                  class="router-link"
                  href="/lists/id3"
                >
                  <h5
                    class="mb-1"
                  >
                    <i
                      class="fa fa-music text-secondary me-3"
                    />
                    baz
                  </h5>
                </a>
              </div>
              <div
                class="col-lg-4"
              >
                <small
                  class="text-muted"
                >
                  May 31 2020, 12:00:00 am
                </small>
              </div>
              <div
                class=" pe-lg-3 col-lg-2"
              >
                <div
                  class="float-end btn-group"
                  role="group"
                >
                  <button
                    class="p-0 me-3 btn btn-link"
                    data-test-id="incomplete-list-complete"
                    type="button"
                  >
                    <i
                      class="fa fa-check fa-2x text-success"
                    />
                  </button>
                  <a
                    class="p-0 me-3 list-button-enabled"
                    data-test-id="incomplete-list-share"
                    href="/lists/id3/users_lists"
                  >
                    <i
                      class="fa fa-users fa-2x text-primary"
                    />
                  </a>
                  <a
                    class="p-0 mr-3 list-button-enabled"
                    data-test-id="incomplete-list-edit"
                    href="/lists/id3/edit"
                  >
                    <i
                      class="fa fa-edit fa-2x text-warning"
                    />
                  </a>
                  <button
                    class="p-0 btn btn-link"
                    data-test-id="incomplete-list-trash"
                    type="button"
                  >
                    <i
                      class="fa fa-trash fa-2x text-danger"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="accepted-list list-list-group-item list-group-item"
        data-test-class="incomplete-list"
        data-test-id="list-id5"
      >
        <div
          class="row"
        >
          <div
            class="col-sm-12 col-12"
          >
            <div
              class=" pt-1 row"
            >
              <div
                class="col-lg-6"
              >
                <a
                  class="router-link"
                  href="/lists/id5"
                >
                  <h5
                    class="mb-1"
                  >
                    <i
                      class="fa fa-list text-secondary me-3"
                    />
                    foobar
                  </h5>
                </a>
              </div>
              <div
                class="col-lg-4"
              >
                <small
                  class="text-muted"
                >
                  May 31 2020, 12:00:00 am
                </small>
              </div>
              <div
                class=" pe-lg-3 col-lg-2"
              >
                <div
                  class="float-end btn-group"
                  role="group"
                >
                  <button
                    class="p-0 me-3 btn btn-link"
                    data-test-id="incomplete-list-complete"
                    type="button"
                  >
                    <i
                      class="fa fa-check fa-2x text-success"
                    />
                  </button>
                  <a
                    class="p-0 me-3 list-button-disabled"
                    data-test-id="incomplete-list-share"
                    disabled=""
                    href="/lists/id5/users_lists"
                  >
                    <i
                      class="fa fa-users fa-2x text-primary"
                    />
                  </a>
                  <a
                    class="p-0 mr-3 list-button-enabled"
                    data-test-id="incomplete-list-edit"
                    href="/lists/id5/edit"
                  >
                    <i
                      class="fa fa-edit fa-2x text-warning"
                    />
                  </a>
                  <button
                    class="p-0 btn btn-link"
                    data-test-id="incomplete-list-trash"
                    type="button"
                  >
                    <i
                      class="fa fa-trash fa-2x text-danger"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="mb-4"
  >
    <div
      class="clearfix"
    >
      <div
        class="mx-auto float-start"
      >
        <div
          class="m-0 row"
        >
          <h2
            class="text-capitalize pe-0 w-auto"
          >
            Completed
          </h2>
          <i
            class="far fa-question-circle float-end text-secondary ps-0 w-auto"
            data-test-id="Completed-popover"
          />
        </div>
      </div>
      <button
        class="mx-auto float-end btn btn-link"
        type="button"
      >
        
         Select
      </button>
    </div>
    <div
      class="list-group"
    >
      <div
        class="accepted-list list-list-group-item list-group-item"
        data-test-class="completed-list"
        data-test-id="list-id2"
      >
        <div
          class="row"
        >
          <div
            class="col-sm-12 col-12"
          >
            <div
              class=" pt-1 row"
            >
              <div
                class="col-lg-6"
              >
                <a
                  class="router-link"
                  href="/lists/id2"
                >
                  <h5
                    class="mb-1"
                  >
                    <i
                      class="fa fa-book text-secondary me-3"
                    />
                    bar
                  </h5>
                </a>
              </div>
              <div
                class="col-lg-4"
              >
                <small
                  class="text-muted"
                >
                  May 31 2020, 12:00:00 am
                </small>
              </div>
              <div
                class=" pe-lg-3 col-lg-2"
              >
                <div
                  class="float-end btn-group"
                  role="group"
                >
                  <button
                    class="p-0 me-3 list-button-enabled btn btn-link"
                    data-test-id="complete-list-refresh"
                    type="button"
                  >
                    <i
                      class="fa fa-redo fa-2x text-primary"
                    />
                  </button>
                  <button
                    class="p-0 btn btn-link"
                    data-test-id="complete-list-trash"
                    type="button"
                  >
                    <i
                      class="fa fa-trash fa-2x text-danger"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="accepted-list list-list-group-item list-group-item"
        data-test-class="completed-list"
        data-test-id="list-id4"
      >
        <div
          class="row"
        >
          <div
            class="col-sm-12 col-12"
          >
            <div
              class=" pt-1 row"
            >
              <div
                class="col-lg-6"
              >
                <a
                  class="router-link"
                  href="/lists/id4"
                >
                  <h5
                    class="mb-1"
                  >
                    <i
                      class="fa fa-book text-secondary me-3"
                    />
                    bar
                  </h5>
                </a>
              </div>
              <div
                class="col-lg-4"
              >
                <small
                  class="text-muted"
                >
                  May 31 2020, 12:00:00 am
                </small>
              </div>
              <div
                class=" pe-lg-3 col-lg-2"
              >
                <div
                  class="float-end btn-group"
                  role="group"
                >
                  <button
                    class="p-0 me-3 list-button-disabled btn btn-link"
                    data-test-id="complete-list-refresh"
                    disabled=""
                    type="button"
                  >
                    <i
                      class="fa fa-redo fa-2x text-primary"
                    />
                  </button>
                  <button
                    class="p-0 btn btn-link"
                    data-test-id="complete-list-trash"
                    type="button"
                  >
                    <i
                      class="fa fa-trash fa-2x text-danger"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
